bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/BundleProcessorVetoManager.java

```diff
@@ -0,0 +1,146 @@
+
+    private final Set<BundleProcessor> bundleProcessors = new HashSet<BundleProcessor>();
+    private final Multimap<Long, BundleProcessor> vetoes = Multimaps
+            .synchronizedListMultimap(LinkedListMultimap.<Long, BundleProcessor> create());
+    private final Multimap<Long, T> queue = Multimaps.synchronizedListMultimap(LinkedListMultimap.<Long, T> create());
+    private final Action<T> action;
+
+    /**
+     *
+     * @param action the action to run (potentially delayed)
+     */
+    public BundleProcessorVetoManager(Action<T> action) {
+        this.action = action;
+    }
+
+    @Override
+    public void bundleFinished(final BundleProcessor context, final Bundle bundle) {
+        vetoes.remove(bundle.getBundleId(), context);
+        if (vetoes.get(bundle.getBundleId()).isEmpty()) {
+            logger.debug("Finished loading meta-data of bundle '{}'.", bundle.getSymbolicName());
+            synchronized (queue) {
+                for (T object : queue.removeAll(bundle.getBundleId())) {
+                    action.apply(object);
+                }
+            }
+        } else {
+            logger.debug("'{}' still vetoed by '{}', queueing '{}'", bundle.getSymbolicName(),
+                    vetoes.get(bundle.getBundleId()), queue.get(bundle.getBundleId()));
+        }
+    }
+
+    private Bundle getBundle(final Class<?> classFromBundle) {
+        final ClassLoader classLoader = AccessController.doPrivileged(new PrivilegedAction<ClassLoader>() {
+            @Override
+            public ClassLoader run() {
+                return classFromBundle.getClassLoader();
+            }
+        });
+        if (classLoader instanceof BundleReference) {
+            Bundle bundle = ((BundleReference) classLoader).getBundle();
+            logger.trace("Bundle of {} is {}", classFromBundle, bundle.getSymbolicName());
+            return bundle;
+        }
+        return null;
+    }
+
+    /**
+     * Run the given action.
+     *
+     * If necessary, it will get queued and executed once the binding's XML files have been loaded.
+     *
+     * @param object the argument for the action
+     */
+    public void applyActionFor(final T object) {
+        boolean veto = false;
+        Bundle bundle = getBundle(object.getClass());
+        long bundleId = bundle.getBundleId();
+        for (BundleProcessor proc : bundleProcessors) {
+            if (!proc.hasFinishedLoading(bundle)) {
+                veto = true;
+                if (!vetoes.containsEntry(bundleId, proc)) {
+                    logger.trace("Marking '{}' as vetoed by '{}'", bundle.getSymbolicName(), proc);
+                    vetoes.put(bundleId, proc);
+                }
+            } else {
+                logger.trace("'{}' already finished processing '{}'", proc, bundle.getSymbolicName());
+            }
+        }
+        if (veto) {
+            if (!queue.containsEntry(bundle, object)) {
```

Comment:
To be honest, I think we can drop the synchronization here entirely. The collection is synchronized itself and the iteration above if over a collection of removed elements, which I don't expect it can change.